ssstar/src/objstore/mod.rs

```diff
@@ -142,6 +180,43 @@ pub(crate) trait Bucket: DynClone + std::fmt::Debug + Sync + Send + 'static {
 
 dyn_clone::clone_trait_object!(Bucket);
 
+/// Multi-part upload client which provides a high-level API for uploading a large object with a
+/// known size to object storage in multiple parts.
+///
+/// Clones all share the same internal state, so multiple clones can be used to call `parts` and
+/// `upload_part` and the result will be as if all of those calls were made on a single instance.
+///
+/// `finish` must only be called once across all clones or an error ocurrs.
+#[async_trait::async_trait]
+pub(crate) trait MultipartUploader: DynClone + Sync + Send + 'static {
+    /// Initialize this multi-part upload.
+    ///
+    /// This must be called exactly once, and must be the first call made on this object.
+    /// `upload_part` and `finish` will panic if `init` isn't called first.
+    async fn init(&self) -> Result<()>;
```

Comment:
It is likely logical to use the **State machine** design pattern. If so, the caller will never use the methods in an incorrect order.